### PR TITLE
dimaru-coti/rejected_transactions_removal

### PR DESCRIPTION
**Pull request by dimaru-coti**

### CODE REVIEW

File: basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java [hunk 1]
```diff
@@ -105,4 +105,29 @@ public ResponseEntity<IResponse> setFeeValue(@Valid @RequestBody RatioTokenFeeSe
     public ResponseEntity<IResponse> deleteFeeValue(@Valid @RequestBody DeleteTokenFeeRequest deleteTokenFeeRequest) {
         return nodeFeesService.deleteFeeValue(deleteTokenFeeRequest);
     }
+
+    @Operation(summary = "Retrieve List of Rejected Transactions by Admin", operationId = "getRejectedTransactions")
```
Comment (gmesika-coti) [inline basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java:109]: why did you create this method? we have the batch method already

File: basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java [hunk 2]
```diff
@@ -105,4 +105,29 @@ public ResponseEntity<IResponse> setFeeValue(@Valid @RequestBody RatioTokenFeeSe
...
+                    content = {@Content(mediaType = "application/json",
+                            schema = @Schema(implementation = Response.class))})})
+    @GetMapping(path = "/transaction/rejected")
+    public ResponseEntity<IResponse> getRejectedTransactions() {
+        return transactionService.getRejectedTransactions();
+    }
+
+    @Operation(summary = "Delete Rejected Transactions", operationId = "deleteRejectedTransactions")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "List of Deleted Rejected Transactions",
```
Comment (gmesika-coti) [inline basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java]: description is wrong

File: basenode/src/main/java/io/coti/basenode/services/BaseNodeTransactionService.java [hunk 3]
```diff
@@ -37,10 +38,13 @@ public class BaseNodeTransactionService implements ITransactionService {
 
     protected Map<TransactionData, Boolean> postponedTransactionMap = new ConcurrentHashMap<>();  // true/false means new from full node or propagated transaction
     private final LockData transactionLockData = new LockData();
-    protected static final long REJECTED_TRANSACTIONS_TTL = Duration.ofDays(30).getSeconds();
+    protected long rejectedTransactionsTtl;
+    @Value("${number.of.days.ttl:30}")
```
Comment (gmesika-coti) [inline basenode/src/main/java/io/coti/basenode/services/BaseNodeTransactionService.java]: property name not clear . number of days ttl of what? maybe rejected.transaction.ttl ?

File: fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java [hunk 4]
```diff
@@ -765,4 +765,45 @@ public void getAddressRejectedTransactionBatch(GetAddressTransactionBatchRequest
...
+            return ResponseEntity
+                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
+                    .body(new Response(
+                            TRANSACTION_REJECTED_SERVER_ERROR,
+                            STATUS_ERROR));
+        }
+    }
+
+    @Override
+    public ResponseEntity<IResponse> deleteRejectedTransactions(DeleteRejectedTransactionsRequest deleteRejectedTransactionsRequest) {
```
Comment (gmesika-coti) [inline fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java]: code should not be duplicated. can move to TransactionHelper

File: dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java [hunk 5]
```diff
@@ -162,7 +162,31 @@ public ResponseEntity<IResponse> getRejectedTransactions() {
...
                     .status(HttpStatus.INTERNAL_SERVER_ERROR)
                     .body(new Response(
-                            TRANSACTION_NONE_INDEXED_SERVER_ERROR,
+                            TRANSACTION_REJECTED_SERVER_ERROR,
+                            STATUS_ERROR));
+        }
+    }
+
+    @Override
+    public ResponseEntity<IResponse> deleteRejectedTransactions(DeleteRejectedTransactionsRequest deleteRejectedTransactionsRequest) {
```
Comment (gmesika-coti) [inline dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java]: code duplicated. can move to TransactionHelper

File: dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java [hunk 6]
```diff
@@ -162,7 +162,31 @@ public ResponseEntity<IResponse> getRejectedTransactions() {
...
+        try {
+            List<RejectedTransactionResponseData> rejectedTransactionDataList = new ArrayList<>();
+            deleteRejectedTransactionsRequest.getTransactionHashes().forEach(rejectedTransactionHash -> {
+                RejectedTransactionData rejectedTransaction = rejectedTransactions.getByHash(rejectedTransactionHash);
+                if (rejectedTransaction != null) {
```
Comment (gmesika-coti) [inline dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java]: should we do the process if some of the input is wrong?

**Review by gmesika-coti — CHANGES_REQUESTED** [review]

File: basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java (hunk 1 above)
Comment (dimaru-coti) [reply basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java:109]: I have moved it from the DSP node to base as we have discussed that we need it also for Fullnode

File: basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java (hunk 2 above)
Comment (dimaru-coti) [reply basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java]: Why it is not correct?

File: fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java (hunk 4 above)
Comment (dimaru-coti) [reply fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java]: It is not completely same, removeAddressFromRejectedTransactionsMap(rejectedTransaction); is not needed in DSP

File: basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java (hunk 2 above)
Comment (dimaru-coti) [reply basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java]: Successfully deleted rejected transactions

File: basenode/src/main/java/io/coti/basenode/services/BaseNodeTransactionService.java (hunk 3 above)
Comment (dimaru-coti) [reply basenode/src/main/java/io/coti/basenode/services/BaseNodeTransactionService.java]: rejected.transactions.days.ttl

File: basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java (hunk 2 above)
Comment (dimaru-coti) [reply basenode/src/main/java/io/coti/basenode/controllers/BaseNodeAdminController.java]: Updated

File: dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java (hunk 5 above)
Comment (dimaru-coti) [reply dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java]: moved to BaseNodeTransactionService

File: fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java (hunk 4 above)
Comment (dimaru-coti) [reply fullnode/src/main/java/io/coti/fullnode/services/TransactionService.java]: moved to BaseNodeTransactionService

File: dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java (hunk 6 above)
Comment (dimaru-coti) [reply dspnode/src/main/java/io/coti/dspnode/services/TransactionService.java]: added validation loop with error response on failure

**Pull request merged**